include/nbl/video/utilities/CArithmeticOps.h

```diff
@@ -0,0 +1,286 @@
+					temporaryStorageOffset[topLevel-1u] = lastElement[++topLevel] = lastElement[topLevel]/workgroupSize;
+				
+				std::exclusive_scan(temporaryStorageOffset,temporaryStorageOffset+sizeof(temporaryStorageOffset)/sizeof(uint32_t),temporaryStorageOffset,0u);
+			}
+            // given already computed tables of lastElement indices per level, number of levels, and storage offsets, tell us total auxillary buffer size needed
+			inline uint32_t getScratchSize(uint32_t ssboAlignment=256u)
+			{
+				uint32_t uint_count = 1u; // workgroup enumerator
+				uint_count += temporaryStorageOffset[MaxLevels/2u-1u]; // last scratch offset
+				uint_count += lastElement[topLevel]+1u; // and its size
+				return core::roundUp<uint32_t>(uint_count*sizeof(uint32_t),ssboAlignment);
+			}
+		};
+        
+        // the default scheduler we provide works as described above in the big documentation block
+		struct SchedulerParameters : nbl::hlsl::scan::DefaultSchedulerParameters_t  // this struct and its methods are also available in HLSL
+		{
+			SchedulerParameters()
+			{
+				std::fill_n(cumulativeWorkgroupCount,Parameters::MaxLevels,0u);
+				std::fill_n(workgroupFinishFlagsOffset, Parameters::MaxLevels, 0u);
+				std::fill_n(lastWorkgroupSetCountForLevel,Parameters::MaxLevels,0u);
+			}
+            
+            // given the number of elements and workgroup size, figure out how many atomics we need
+            // also account for the fact that we will want to use the same scratch buffer both for the
+            // scheduler's atomics and the aux data storage
+			SchedulerParameters(Parameters& outScanParams, const uint32_t _elementCount, const uint32_t workgroupSize) : SchedulerParameters()
+			{
+				outScanParams = Parameters(_elementCount,workgroupSize);
+				const auto topLevel = outScanParams.topLevel;
+
+				std::copy_n(outScanParams.lastElement+1u,topLevel,cumulativeWorkgroupCount);
+				for (auto i=0u; i<=topLevel; i++)
+					cumulativeWorkgroupCount[i] += 1u;
+				std::reverse_copy(cumulativeWorkgroupCount,cumulativeWorkgroupCount+topLevel,cumulativeWorkgroupCount+topLevel+1u);
+
+				for (auto i = 0u; i <= topLevel; i++) {
+					workgroupFinishFlagsOffset[i] = ((cumulativeWorkgroupCount[i] - 1u) >> hlsl::findMSB(workgroupSize - 1u)) + 1;
+					lastWorkgroupSetCountForLevel[i] = cumulativeWorkgroupCount[i] & (workgroupSize - 1u);
+				}
+
+				const auto wgFinishedFlagsSize = std::accumulate(workgroupFinishFlagsOffset, workgroupFinishFlagsOffset + Parameters::MaxLevels, 0u);
+				std::exclusive_scan(workgroupFinishFlagsOffset, workgroupFinishFlagsOffset + Parameters::MaxLevels, workgroupFinishFlagsOffset, 0u);
+				for (auto i=0u; i<sizeof(Parameters::temporaryStorageOffset)/sizeof(uint32_t); i++)
+					outScanParams.temporaryStorageOffset[i] += wgFinishedFlagsSize;
+				
+				std::inclusive_scan(cumulativeWorkgroupCount, cumulativeWorkgroupCount + Parameters::MaxLevels, cumulativeWorkgroupCount);
```

Comment:
you can add equivalents of the `std::` functions to `algorithm.hlsl`, they'll just be single invocation utilities